[MPS] Extend atomic operations to all int types

malfet opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #158179
* #158178
* #158064

That fixes `index_put(..., accumulate=True)` for all dtypes

int64 operation is not really atomic, but eventually consistent from the `index_put_accumulate` kernel point of view: i.e. by the end of the operation results in the global memory are indeed accumulation of the operands at given indices

malfet committed:
[MPS] Extend atomic operations to more int types

The only ones that are not covered right now are int64

[ghstack-poisoned]

Skylion007 commented on c10/metal/atomic.h:
```diff
@@ -35,15 +35,16 @@ static inline void atomic_add_helper(
   union {
     uint i;
-    T t[2];
+    T t[elem_per_enum];
```
Nit can this be a real array type since array is constexpr length?

Skylion007 replied:
metal::array

malfet committed:
Update on "[MPS] Extend atomic operations to more int types"


That fixes index_put for all dtypes but int64 (as Metal only has atomic loads and stores for this dtype)

[ghstack-poisoned]

malfet commented:
@pytorchbot merge -f "Lint + MPS are green"
